File: core/src/main/scala/stasis/core/security/oauth/OAuthClient.scala
package stasis.core.security.oauth

import stasis.core.security.oauth.OAuthClient.AccessTokenResponse

import scala.concurrent.Future

trait OAuthClient {
  def token(scope: Option[String], parameters: OAuthClient.GrantParameters): Future[AccessTokenResponse]
}

object OAuthClient {
  import play.api.libs.json.{Format, Json}

  object GrantType {
    final val ClientCredentials: String = "client_credentials"
    final val ResourceOwnerPasswordCredentials: String = "password"
    final val RefreshToken: String = "refresh_token"
  }

  sealed trait GrantParameters
  object GrantParameters {
    final case class ClientCredentials() extends GrantParameters
    final case class ResourceOwnerPasswordCredentials(username: String, password: String) extends GrantParameters
    final case class RefreshToken(refreshToken: String) extends GrantParameters
  }

  implicit val accessTokenResponseFormat: Format[AccessTokenResponse] = Json.format[AccessTokenResponse]

  final case class AccessTokenResponse(
    access_token: String,
    refresh_token: Option[String],
    expires_in: Long,
    scope: Option[String]
  )
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/security/OAuthClient.kt
package stasis.client_android.lib.security

import stasis.client_android.lib.utils.Try

interface OAuthClient {
    suspend fun token(scope: String?, parameters: GrantParameters): Try<AccessTokenResponse>

    object GrantType {
        const val ClientCredentials: String = "client_credentials"
        const val ResourceOwnerPasswordCredentials: String = "password"
        const val RefreshToken: String = "refresh_token"
    }

    sealed class GrantParameters {
        object ClientCredentials : GrantParameters()
        data class ResourceOwnerPasswordCredentials(val username: String, val password: String) : GrantParameters()
        data class RefreshToken(val refreshToken: String) : GrantParameters()
    }
}
